hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSImageFormatPBINode.java

```diff
@@ -217,33 +272,147 @@ void loadINodeDirectorySection(InputStream in) throws IOException {
         INodeDirectory p = dir.getInode(e.getParent()).asDirectory();
         for (long id : e.getChildrenList()) {
           INode child = dir.getInode(id);
-          addToParent(p, child);
+          if (addToParent(p, child)) {
+            if (child.isFile()) {
+              inodeList.add(child);
+            }
+            if (inodeList.size() >= 1000) {
+              addToCacheAndBlockMap(inodeList);
+              inodeList.clear();
+            }
+          }
```

Comment:
better to log a warning if addToParent() returns false. Otherwise inode may be dropped silently.

Reply:
I have added a message like this to both adding the inode and inode references to the directory:
```
LOG.warn("Failed to add the inode reference {} to the directory {}",
                ref.getId(), p.getId());
```
I opted to log only the inode and directory "inode id" as I am not sure if the system will be able to resolve the full path of an inode or directory at this stage, as it is still loading the image. Also this "should never happen" so hopefully we will not see these messages in practice, but if we do, it will likely require manually investigating an image corruption, so the ID numbers should be enough to start with that.